File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    const l = words.length
    const visited = new Array(l).fill(false);
    const q = [[begin,0]];
    let idx = 0;
    while(idx < q.length){
        const [cur,cI] = q[idx++];
        if(cur === target) return cI;
        for(let i = 0; i < l; i++){
            let diff = 0;
            if(!visited[i]){
                const candidate = words[i];
                for(let j = 0; j < begin.length; j++){
                    if(cur[j] !== candidate[j]) diff++;
                }
                if(diff === 1) {
                    visited[i] = true
                    q.push([candidate, cI + 1]);
                }
            }
        }
    }
    return 0;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    n = len(words)
    word_len = len(begin)
    visited = [False] * n  
    cnt = 0
    q = deque()
    q.append((begin,0))
    
    def is_one_diff(s1, s2):
        return sum(a != b for a, b in zip(s1, s2)) == 1
        
    while q:
        cur, cnt = q.popleft()
        if cur == target:
            return cnt
        
        for i in range(n):
            if visited[i]:
                continue
            if is_one_diff(cur, words[i]):
                q.append((words[i],cnt+1))
                visited[i] = True
    return 0